### DISCUSSION TITLE
Reading clipboard in a custom event handler

### DISCUSSION THREAD

**Discussion by tpoliaw** [wezterm/wezterm#4425] [Q&A] [answered]
### What Operating System(s) are you running on?

Linux X11

### Which Wayland compositor or X11 Window manager(s) are you using?

i3 version 4.22 (2023-01-02) © 2009 Michael Stapelberg and contributors

### WezTerm version

wezterm 20231002-222603-79ae543b

### Ask your question!

I am trying to temporarily work around the paste bug/issue [here](https://github.com/wez/wezterm/issues/4402).

In my config I have added what I thought would be the relevant parts

```lua
function custom_paste(window, pane)
   handle = io.popen("xclip -o")
   text = handle:read("*a")
   handle:close()
   window:perform_action(
      wezterm.action.SendString(text),
      pane
   )
end

wezterm.on("custom-paste", custom_paste)

config = {
  keys = {
    { key = "Insert", mods = "SHIFT", action = wezterm.action.EmitEvent('custom-paste') }
  },
}
```
<details>
  <summary>Error from wezterm log</summary>
   <pre>
  12:04:15.621  ERROR  wezterm_gui                    > running message loop: process_queued_xcb: ChangeProperty {
    mode: Replace,
    window: Window {
        res_id: 37748736,
    },
    property: Atom {
        res_id: 961,
    },
    type: Atom {
        res_id: 314,
    },
    data: [
        49,
        45,
        111,
        114,
        103,
        46,
        119,
        101,
        122,
        102,
        117,
        114,
        108,
        111,
        110,
        103,
        46,
        119,
        101,
        122,
        116,
        101,
        114,
        109,
    ],
}: X(Window(ValueError { response_type: 0, error_code: 3, sequence: 1744, bad_value: 37748736, minor_opcode: 0, major_opcode: 18, pad: 1 }), Some("x::ChangeProperty")); terminating
   </pre>
  ```
</details>

This works fine for the original use case that prompted the issue as well as copying from external applications. When I try to select/paste from wezterm it crashes and kill all instances. Even if I remove the `perform_action` call so it's not trying to update the GUI, it still crashes so I'm guessing at this point wezterm is in a state where it can't respond the the X request for the clipboard data so X is timing out or wezterm is avoiding a deadlock by crashing? Is there a better way to approach this where wezterm won't be blocking itself?

I've now spent a while digging into how X clipboards work so I might open a PR for the original problem but until then I thought this might be quick way to get something working.

**Comment by wez** [comment, marked as answer]
I don't have time to look at this right now, but after a quick skim: Don't use `io.popen` in wezterm; reading from a pipe will block, and since reading the clipboard requires co-operative IPC between the relevant processes. I wouldn't be surprised to see some kind of deadlock or error situation arise from using it.

Instead, consider using https://wezfurlong.org/wezterm/config/lua/wezterm/run_child_process.html which will run asynchronously.

**Reply by tpoliaw** [reply]
Thanks, that sounds ideal. I'll give it a go.
